File: js/entities/player.js
// Player Entity Module

import { Physics } from '../engine/physics.js';

export class Player {
    constructor(id, name, role = 'attacker', isHuman = false, number = 1, color = '#00ff87', charData = {}) {
        this.id = id;
        this.name = name;
        this.role = role; // 'attacker' or 'defender'
        this.isHuman = isHuman;
        this.number = number;
        this.color = color;
        
        // Cartoon Character Attributes
        this.skinColor = charData.skinColor || '#f5cda7';
        this.hairColor = charData.hairColor || '#1a1a1a';
        this.jerseyColor = charData.jerseyColor || color;
        this.secondaryColor = charData.secondaryColor || '#ffffff';
        this.stripeColor = charData.stripeColor || null;
        this.beard = charData.beard || false;
        this.headband = charData.headband || null;
        this.ponytail = charData.ponytail || false;
        this.armband = charData.armband || null;
        this.style = charData.style || 'default';
        
        this.x = 0;
        this.y = 0;
        this.vx = 0;
        this.vy = 0;
        
        this.radius = 20; // Increased radius for clear cartoon avatar details
        this.baseSpeed = role === 'defender' ? 3.4 : 3.0;
        this.facingAngle = 0;
        
        // Designated home slot on perimeter (for attackers)
        this.baseAngle = 0;
        this.targetX = 0;
        this.targetY = 0;
        
        // Pass selection state
        this.selectedTarget = null;
        this.passCooldown = 0;
        this.hasBall = false;
        
        // AI State
        this.aiShiftAngle = 0;
    }

    setPosition(x, y) {
        this.x = x;
        this.y = y;
        this.targetX = x;
        this.targetY = y;
    }

    update(pitchCenter, rondoRadius) {
        // Cooldown timer
        if (this.passCooldown > 0) this.passCooldown--;

        // Move towards target position smoothly
        const dx = this.targetX - this.x;
        const dy = this.targetY - this.y;
        const dist = Math.sqrt(dx * dx + dy * dy);
        
        if (dist > 1) {
            const moveSpeed = Math.min(this.baseSpeed, dist * 0.15);
            this.vx = (dx / dist) * moveSpeed;
            this.vy = (dy / dist) * moveSpeed;
            this.x += this.vx;
            this.y += this.vy;
        } else {
            this.vx = 0;
            this.vy = 0;
        }

        // Clamp to field perimeter if attacker
        if (this.role === 'attacker') {
            const distFromCenter = Physics.distance(this, pitchCenter);
            if (Math.abs(distFromCenter - rondoRadius) > 40) {
                const angle = Physics.angleBetween(pitchCenter, this);
                this.x = pitchCenter.x + Math.cos(angle) * rondoRadius;
                this.y = pitchCenter.y + Math.sin(angle) * rondoRadius;
            }
        }
    }
}

File: lib/models/player.dart
import 'dart:math';
import 'package:flutter/material.dart';
import '../engine/physics.dart';

class Player {
  final int id;
  final String name;
  final String role; // 'attacker' or 'defender'
  bool isHuman;
  final int number;
  final Color jerseyColor;
  final Color secondaryColor;
  final Color hairColor;
  final Color skinColor;
  final Color? stripeColor;
  final bool beard;
  final Color? headband;
  final bool ponytail;
  final Color? armband;
  final String style;

  double x = 0;
  double y = 0;
  double vx = 0;
  double vy = 0;

  double radius = 20.0;
  double baseSpeed = 3.0;
  double facingAngle = 0.0;

  double baseAngle = 0.0;
  double targetX = 0.0;
  double targetY = 0.0;

  int passCooldown = 0;
  bool hasBall = false;

  Player({
    required this.id,
    required this.name,
    this.role = 'attacker',
    this.isHuman = false,
    required this.number,
    required this.jerseyColor,
    this.secondaryColor = Colors.white,
    this.hairColor = const Color(0xFF1A1A1A),
    this.skinColor = const Color(0xFFF5CDA7),
    this.stripeColor,
    this.beard = false,
    this.headband,
    this.ponytail = false,
    this.armband,
    this.style = 'default',
  }) {
    baseSpeed = role == 'defender' ? 3.4 : 3.0;
  }

  void setPosition(double newX, double newY) {
    x = newX;
    y = newY;
    targetX = newX;
    targetY = newY;
  }

  void update(double pitchCenterX, double pitchCenterY, double rondoRadius) {
    if (passCooldown > 0) passCooldown--;

    final dx = targetX - x;
    final dy = targetY - y;
    final dist = sqrt(dx * dx + dy * dy);

    if (dist > 1.0) {
      final moveSpeed = min(baseSpeed, dist * 0.15);
      vx = (dx / dist) * moveSpeed;
      vy = (dy / dist) * moveSpeed;
      x += vx;
      y += vy;
    } else {
      vx = 0;
      vy = 0;
    }

    if (role == 'attacker') {
      final distFromCenter = Physics.distance(x, y, pitchCenterX, pitchCenterY);
      if ((distFromCenter - rondoRadius).abs() > 40.0) {
        final angle = Physics.angleBetween(pitchCenterX, pitchCenterY, x, y);
        x = pitchCenterX + cos(angle) * rondoRadius;
        y = pitchCenterY + sin(angle) * rondoRadius;
      }
    }
  }
}
